### PR TITLE
At ListMatch syntax to promql

### PR DESCRIPTION
**Pull request by jacksontj**
In usage of promql its quite common to want to target a specific label by a known set of values (the canonical example being the `node` label with a list of known hostnames). In the current promql language the only mechanism you have to do a multi-select is to use a regex which would look something like:

`metric_name{node=~"(host1|host1)"}`

The downside of this being that I have to take my nicely structured data (a list) and put it into a regex. This causes a variety of performance problems. There is an issue up right now (#2651) to "optimize queries using regex matchers for set lookups". In my experience writing code to parse things out of regexes to do smarter things is *significantly* harder than just keeping the data structured to begin with. Although there may be other use-cases that would want "smarter regex" the "match a list of values" seems like a common-enough case to warrant a new matcher.

I have (in my fork) new matchers to do a valueSet comparison. With this feature the above query becomes:

`metric_name{node=|"host1,host1"}`

I originally was planning on doing `metric_name{node=["host1","host1"]}` but that seemed to cause quite the pain with the lexer/parser (as there aren't any other matchers with trailing characters.

I have the change staged in my fork of tsdb (https://github.com/prometheus/tsdb/compare/master...jacksontj:master) before this would be merged I'll get that in and then update the vendor normally.

### CODE REVIEW

**Comment by brian-brazil** [comment]
Thank you for your PR. I'm not sure this makes sense to add to PromQL, as we already support this via regex and adding another syntax is likely to cause confusion in addition to requiring more work on the part of remote read endpoints. I think our current plan of handling this via a backend optimisation is way to go, rather than requiring the user to care.

**Comment by jacksontj** [comment]
@brian-brazil I definitely think going from specific to general is significantly better than general to specific.

Applying that methodology to this problem, it is significantly easier if we have more explicit query mechanisms that get pushed into more general ones.

As an example lets say we implemented the following:
(1) SimpleStorage interface: this includes exact and regex
(2) AdvancedStorage interface: this includes exact, regex, lists, etc.

In the server side we can easily convert anything from AdvancedStorage to SimpleStorage, an example changing a query of `metric_name{node=|"host1,host1"}` -> `metric_name{node=~"(host1|host1)"}` is 100% possible.

With this mechanism we still can allow users to define queries as explicitly as they want, but retain the "simple" interface with the storage systems.

If we follow your suggestion of having the storage nodes attempt to introspect the regex, we'll end up creating a large mess. When implementing this "reverse a regex" logic you can start out with specific hard-coded cases, but this quickly becomes an unsolveable problem (as regex lets you do ALL sorts of magic).

The specificity of the query becomes even *more* important when you get to federation etc. as a regex requires sending the query to every possible location  whereas if its a list for example you could potentially route the query to the correct prometheus cluster.

TLDR; Its significantly easier to go specific -> general, and usually (almost) impossible to go from general -> specific.
